File: server/src/routes/recommendations.ts
// SPDX-License-Identifier: MIT
// Port of py/api/routes/recommendations.py
// Route: POST /api/v1/recommendations

import { Hono } from 'hono';
import { z } from 'zod';
import type { RecommendService } from '../services/recommendService.js';
import { TYPE_NAMES } from '../types.js';

const RecommendationRequestSchema = z.object({
  type_counts: z.record(z.string(), z.number().int().min(0)).optional(),
  no_support: z.boolean().default(false),
  borrow_friend: z.boolean().optional().default(false),
});

export function createRecommendationsRouter(recommendService: RecommendService): Hono {
  const router = new Hono();

  router.post('/', async (c) => {
    let body: z.infer<typeof RecommendationRequestSchema>;
    try {
      body = RecommendationRequestSchema.parse(await c.req.json());
    } catch (err) {
      return c.json({ error: String(err) }, 400);
    }

    const { type_counts, no_support, borrow_friend } = body;

    let result;
    if (!type_counts || Object.values(type_counts).every((n) => n === 0)) {
      // No type distribution specified — return best 6 overall
      result = await recommendService.getBestCards(no_support);
    } else {
      // Convert string keys to numbers (JSON object keys are always strings)
      const numericTypeCounts: Record<number, number> = {};
      for (const [k, v] of Object.entries(type_counts)) {
        numericTypeCounts[parseInt(k, 10)] = v;
      }
      result = await recommendService.getRecommendedDeck(
        numericTypeCounts,
        no_support,
        borrow_friend ?? false,
      );
    }

    const cards = [
      ...(result.support
        ? [
            {
              name: result.support.name,
              type: result.support.type,
              type_name: TYPE_NAMES[result.support.type] ?? 'Unknown',
              score: result.support.score,
              tier: result.support.tier,
              lb: null,
              rarity: null,
              is_borrowed: true,
            },
          ]
        : []),
      ...result.selected.map((card) => ({
        name: card.name,
        type: card.type,
        type_name: TYPE_NAMES[card.type] ?? 'Unknown',
        score: card.score,
        tier: card.tier,
        lb: card.lb,
        rarity: card.rarity,
        is_borrowed: false,
      })),
    ];

    const totalScore = cards.reduce((s, c) => s + (c.score ?? 0), 0);

    return c.json({ cards, total_score: totalScore });
  });

  return router;
}

File: py/api/routes/recommendations.py
"""API routes for deck recommendations."""

from fastapi import APIRouter, HTTPException, status
from pathlib import Path

from ..models.card import RecommendationRequest, RecommendationResponse, RecommendedCard
from ..services.recommendation_service import RecommendationService

router = APIRouter(prefix="/recommendations", tags=["recommendations"])

# Initialize service
DATA_DIR = Path(__file__).resolve().parent.parent.parent.parent
recommendation_service = RecommendationService(DATA_DIR)


@router.post("", response_model=RecommendationResponse)
async def get_recommendations(request: RecommendationRequest):
    """Get deck recommendations based on type distribution or just best cards.

    Returns a 6-card deck consisting of:
    - 5 cards from your collection (or best available if type_counts is empty)
    - 1 borrowed card from the tierlist (unless no_support=True)

    Note: The game requires exactly one borrowed card in every deck.
    """
    try:
        # If no type counts specified, return best cards
        if not request.type_counts or sum(request.type_counts.values()) == 0:
            selected, support_card = recommendation_service.get_best_cards(
                no_support=request.no_support
            )
        else:
            selected, support_card = recommendation_service.get_recommended_deck(
                type_counts=request.type_counts,
                no_support=request.no_support,
                borrow_friend=request.borrow_friend
            )

        # Convert to response format
        recommended_cards = []

        # Add selected cards from collection FIRST
        for card in selected:
            recommended_cards.append(
                RecommendedCard(
                    name=card.get("name", "Unknown"),
                    type=card.get("type", 0),
                    score=card.get("score", 0),
                    tier=card.get("tier"),
                    lb=card.get("lb"),
                    rarity=card.get("rarity"),
                    is_borrowed=False,
                )
            )

        # Add borrowed card at the BOTTOM if present
        if support_card:
            card_id, name, card_type, score, tier = support_card
            recommended_cards.append(
                RecommendedCard(
                    name=name,
                    type=card_type,
                    score=score,
                    tier=tier,
                    lb=4,  # Borrowed cards are always at MLB
                    rarity=3,  # Show as SSR for display purposes
                    is_borrowed=True,
                )
            )

        # Calculate total score
        total_score = sum(c.score for c in recommended_cards)

        return RecommendationResponse(cards=recommended_cards, total_score=total_score)

    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate recommendations: {str(e)}",
        )
